Add: Template and template_lock to post types endpoint.

jorgefilipecosta opened the pull request:
Backports https://github.com/WordPress/gutenberg/pull/62488 into the core.
Adds template and template_lock property of the post type to post types REST API endpoint.

This change allows us to fix a bug where the template of a page is not respected when creating a new page on the site editor.

Trac ticket:https://core.trac.wordpress.org/ticket/61477

## Testing
With the Gutenberg plugin disabled.
I added the following test code to a PHP file:

```
function myplugin_register_template() {
    $post_type_object = get_post_type_object( 'page' );
    $post_type_object->template = array(
        array( 'core/image' ),
    );
	$post_type_object->template_lock = 'all';
}
add_action( 'init', 'myplugin_register_template' );
```

I opened the site editor and pasted:
wp.data.select('core').getPostType( 'page' );

I verified the template and template_lock information was included.
I repeated the test with the Gutenberg plugin enabled to make sure things worked well together.

jorgefilipecosta committed:
add test cases for template and template_lock (+1 squashed commit)
Squashed commits:
[fe4d5cc905] add test cases for template and template_lock

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -246,6 +246,14 @@ public function prepare_item_for_response( $item, $request ) {
+			$data['template'] = $post_type->template;
+		}
+
+		if ( rest_is_field_included( 'template_lock', $fields ) && ! empty( $post_type->template_lock ) && false !== $post_type->template_lock ) {
```
```suggestion
		if ( rest_is_field_included( 'template_lock', $fields ) && ! empty( $post_type->template_lock ) ) {
```

`empty` already checks for the value being false.

TimothyBJacobs reviewed (changes requested):
We don't typically completely omit properties if they don't have a value defined. If these don't have a natural default, we should set them to `null` and add `null` as an allowed type. See the `icon` property for an example.

This also needs tests.

jorgefilipecosta commented:
> We don't typically completely omit properties if they don't have a value defined. If these don't have a natural default, we should set them to `null` and add `null` as an allowed type. See the `icon` property for an example.
> 
> This also needs tests.

Hi @TimothyBJacobs, thank you for the review 😊

In the case of template_lock which is an enum with the following possible values array( 'all', 'insert', 'contentOnly' ) would it still be ok to return null when a template_lock does not exist?

TimothyBJacobs commented:
Yep! We just need to make sure to add it to the enum list as well.

oandregal commented:
> We don't typically completely omit properties if they don't have a value defined. If these don't have a natural default, we should set them to null and add null as an allowed type. See the icon property for an example.

I quickly checked for this in my review, thinking specifically about the enum. I approved based on that we actually do this in some places: for example, [see the `source` field](https://github.com/WordPress/wordpress-develop/blame/84bef95cdf68f8b020ec4fea395c0715d36b73ae/src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php#L187), which is an enum and is omitted. I'm happy to go either way, but I want to make sure I understand the rationale/preference.

TimothyBJacobs commented:
Gotcha, yeah I don't think we should be doing that there either really. I think for the most part we should be trying to keep the actual properties still defined.

TimothyBJacobs commented:
Is an empty array a valid template? Just want to make sure it's not, otherwise the current code will convert it to `null`. If it is, we can use `isset` instead.

jorgefilipecosta commented:
> Is an empty array a valid template? Just want to make sure it's not, otherwise the current code will convert it to `null`. If it is, we can use `isset` instead.

Hi @TimothyBJacobs, very good point, an empty template is probably not expected but someone may be doing that I replaced the condition with isset.

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php:
```diff
@@ -246,6 +246,14 @@ public function prepare_item_for_response( $item, $request ) {
 		}
 
+		if ( rest_is_field_included( 'template', $fields ) ) {
+			$data['template'] = ! isset( $post_type->template ) ? null : $post_type->template;
```
```suggestion
			$data['template'] = $post_type->template ) ?? null;
```

Can we use here `??`

jorgefilipecosta replied:
Hi end-up using the same logic used at https://github.com/WordPress/wordpress-develop/blob/5c7f94a6aa979dc70ff935e9006957fe0f262932/src/wp-admin/edit-form-blocks.php#L231.

jorgefilipecosta replied:
Yes, I followed this logic 👍

jorgefilipecosta commented:
> > Is an empty array a valid template? Just want to make sure it's not, otherwise the current code will convert it to `null`. If it is, we can use `isset` instead.
> 
> Hi @TimothyBJacobs, very good point, an empty template is probably not expected but someone may be doing that I replaced the condition with isset.


It seems like an empty array is even the default: https://github.com/jorgefilipecosta/wordpress-develop/blob/f11d95acf7e65c86733db21975a652820d4dd1fd/src/wp-includes/class-wp-post-type.php#L548-L549

I'm not too fond of these defaults, but they are already like that for a long time so I updated the API to match them.

mukeshpanchal27 commented on tests/phpunit/tests/rest-api/rest-post-types-controller.php:
```diff
@@ -77,6 +77,24 @@ public function test_get_item_cpt() {
 		$this->check_post_type_object_response( 'view', $response, 'cpt' );
 	}
 
+	public function test_get_item_template_cpt() {
```
`@ticket` annotation is missing.

jorgefilipecosta commented:
Committed at https://github.com/WordPress/wordpress-develop/commit/5d41bdae525945be68b82fd86b35db3693c7cdc9.
